pkg/k8s/deleter/deleter.go

```diff
@@ -0,0 +1,188 @@
+	client  client.Client
+	log     logr.Logger
+}
+
+type objectMetaList []metav1.ObjectMeta
+
+func (ol objectMetaList) asCommaSeparatedString() string {
+	out := ""
+	separator := ""
+	for n := range ol {
+		object := ol[n]
+		out += fmt.Sprintf("%s%s/%s", separator, object.Namespace, object.Name)
+		separator = ", "
+	}
+	return out
+}
+
+func New(log logr.Logger, cluster *v1beta1.Cluster, remoteClient client.Client) Deleter {
+	return Deleter{
+		cluster: cluster,
+		client:  remoteClient,
+		log:     log,
+	}
+}
+
+func (d *Deleter) DeleteServicesWithLoadBalancer(ctx context.Context) error {
+	err := deleteServicesWithLoadBalancer(ctx, d.client, d.log)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func deleteServicesWithLoadBalancer(
+	ctx context.Context,
+	c client.Client,
+	log logr.Logger,
+) error {
+	log.Info("Listing Services with type LoadBalancer")
+	services := &corev1.ServiceList{}
+	err := c.List(ctx, services)
+	if err != nil {
+		return fmt.Errorf("error listing Services: %w", err)
+	}
+
+	svcsFailedToBeDeleted := make(objectMetaList, 0)
+	for i := range services.Items {
+		svc := &services.Items[i]
+		if needsDelete(svc) {
+			log.Info(fmt.Sprintf("Deleting Service %s/%s", svc.Namespace, svc.Name))
+			if err = c.Delete(ctx, svc); err != nil {
+				log.Error(
+					err,
+					fmt.Sprintf(
+						"Error deleting Service %s/%s",
+						svc.Namespace,
+						svc.Name,
+					),
+				)
+				svcsFailedToBeDeleted = append(svcsFailedToBeDeleted, svc.ObjectMeta)
+				continue
+			}
+			if err = waitForServiceDeletion(ctx, c, svc); err != nil {
+				log.Error(
+					err,
+					fmt.Sprintf(
+						"Error waiting for Service to be deleted %s/%s",
+						svc.Namespace,
+						svc.Name,
+					),
+				)
+				svcsFailedToBeDeleted = append(svcsFailedToBeDeleted, svc.ObjectMeta)
+			}
+		}
+	}
+	if len(svcsFailedToBeDeleted) > 0 {
+		return failedToDeleteServicesError(svcsFailedToBeDeleted)
+	}
+
+	return nil
+}
+
+// needsDelete will return true if the Service needs to be deleted to allow for cluster cleanup.
+func needsDelete(service *corev1.Service) bool {
+	if service.Spec.Type != corev1.ServiceTypeLoadBalancer ||
+		len(service.Status.LoadBalancer.Ingress) == 0 {
+		return false
+	}
+	return service.Status.LoadBalancer.Ingress[0].IP != "" ||
+		service.Status.LoadBalancer.Ingress[0].Hostname != ""
+}
+
+func waitForServiceDeletion(
+	ctx context.Context,
+	c client.Client,
+	service *corev1.Service,
+) (err error) {
+	backoff := wait.Backoff{
+		Duration: 1 * time.Second,
+		Factor:   1.5,
+		Jitter:   0,
+		Steps:    13,
+	}
+	// the error is always nil to retry but is captured in the named return err
+	_ = wait.ExponentialBackoff(backoff, func() (bool, error) {
+		key := client.ObjectKey{
+			Namespace: service.Namespace,
+			Name:      service.Name,
+		}
+		err = c.Get(ctx, key, service)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				err = nil
+				return true, nil
+			}
+		}
+		return false, nil
+	})
+
+	return
```

Comment:
Be simple: remove named return (prefer explicit return), capture error as normal, returning nil from func if `errors.IsNotFound` is true.

```suggestion
func waitForServiceDeletion(
	ctx context.Context,
	c client.Client,
	service *corev1.Service,
) (error) {
	backoff := wait.Backoff{
		Duration: 1 * time.Second,
		Factor:   1.5,
		Jitter:   0,
		Steps:    13,
	}
	return wait.ExponentialBackoff(backoff, func() (bool, error) {
		key := client.ObjectKey{
			Namespace: service.Namespace,
			Name:      service.Name,
		}
		err := c.Get(ctx, key, service)
		if err != nil {
			if errors.IsNotFound(err) {
				return true, nil
			}
			return false, err
		}
		return false, nil
	})
```